Introducing "file" events

trygve-lie opened the pull request:
The main idea behind [the](https://github.com/asset-pipe/asset-pipe-sink-s3) [sinks](https://github.com/asset-pipe/asset-pipe-sink-mem) in this project is to provide a common way to plug in different storage types in the [asset-pipe-build-server](https://github.com/asset-pipe/asset-pipe-build-server). Iow, the sinks are intended to provide a common interface that bridge http io in the asset-pipe-build-server to what ever back-end storage we want to support.

The interface as is must be a class that must have a `writer()` method which returns a `WritableStream` and a `reader()` method which returns a `ReadableStream`.

In its simplest form, this makes it possible for us to ex use this sink for writing and reading to the file system as follow in the asset-pipe-build-server when running in development:

```js
const app = express();
const sink = new SinkFs('./someDir');
app.get('/feed/:file', (req, res, next) => {
    const reader = sink.reader(req.params.file);
    reader.pipe(res);
});
```

And then replace the sink with ex [something that writes and reads to Amazon S3](https://github.com/asset-pipe/asset-pipe-sink-s3) in production:

```js
const app = express();
const sink = new SinkS3({s3:'config stuff'});
app.get('/feed/:file', (req, res, next) => {
    const reader = sink.reader(req.params.file);
    reader.pipe(res);
});
```

The above `.pipe()` is very naive and won't really fly in production. Inside the asset-pipe-build-server we need something stronger and we need to handle sending proper http status codes etc. This PR introduces the concept of "file events" in the interface of the sinks. Iow; the sinks should emit a set of events telling us if files exist, does not exist etc.

The following events are introduces in the `.reader()` method:

 - `file found` - Emitted when the wanted file exist in the storage and are ready to be streamed.
 - `file not found` - Emitted when the wanted file does not exist in the storage.

With this we can write a more solid implementation in the asset-pipe-build-server:

```js
const app = express();
const sink = new SinkFs('./someDir');
app.get('/feed/:file', (req, res, next) => {
    const reader = sink.reader(req.params.file);
    reader.on('file not found', () => {
        res.status(404).send('File not found');
    });
    reader.on('file found', () => {
        res.writeHead(200, {});
        reader.pipe(res);
    });
});
```

Having a set of unified events like this is handy since each type of storage will provide info about files existence etc differently.

The question is also; which other events do we need?

This PR does also add proper tests of the `.reader()` method.

sveisvei commented on lib/sink.js:
```diff
@@ -8,52 +8,57 @@ const path = require('path');
-    proxy.pipe(parser).pipe(hasher);
-    proxy.pipe(file);
+        file.on('error', (error) => {
+            console.log(error);
```
This log should not be here imo.

trygve-lie replied:
Nop. This PR only focus on the `.reader()` method.

sveisvei commented on lib/sink.js:
```diff
@@ -8,52 +8,57 @@ const path = require('path');
-    file.on('finish', () => {
-        if (callback) {
-            callback();
+        class ReadStream extends fs.createReadStream {
```
This looks abit weird. Why extend?

trygve-lie replied:
So the original events on the stream (`done`, `error` etc.) is exposed.

SimenB replied:
And why inline?

trygve-lie replied:
No real reason. Can be moved out. Can do when `writer()` gets a rewrite.

sveisvei commented:
The idea of custom events looks good, so the change gets a 👍

SimenB commented on lib/sink.js:
```diff
@@ -8,52 +8,57 @@ const path = require('path');
-    assert(fileDir, '"fileDir" must be provided');
+module.exports = class SinkFs {
+    constructor (fileDir) {
+        assert(fileDir, '"fileDir" is missing');
```
should we throw if `fileDir` is not absolute?

trygve-lie replied:
Does absolute / relative really matter? 

I think we should throw if path does not exist.

SimenB replied:
It doesn't make sense to not pass an absolute path, as the rename you do later will either put things into node_modules, or some other place. Enforcing absolute path with a nice error makes sense, IMO. Not a biggie, though 😄

trygve-lie replied:
Lets do this in a separate task. There is more to this:

 - Temp file should be written to the temp directory of the OS. Not the same directory as the destination of the file.
 - There should be a check which throws if destination directory does not exist.
 - I also think the constructor of the sinks should take a options object instead of an argument like this. This sink is pretty simple, but the config one need to pass in to ex a S3 sink is more complex.

trygve-lie commented:
I've addressed some of the things discussed further up and introduced events in the `writer()` also. The `writer()` does now emit the following events:

* `file not saved` - When it was not possible to save a file to disc.
* `file saved` - When a file is done written to disc.

The constructor is also changed to take a `options` object.
